rllib/evaluation/multi_agent_sample_collector.py

```diff
@@ -0,0 +1,252 @@
+import logging
+from typing import Dict, Optional
+
+from ray.rllib.env.base_env import _DUMMY_AGENT_ID
+from ray.rllib.evaluation.episode import MultiAgentEpisode
+from ray.rllib.evaluation.per_policy_sample_collector import \
+    _PerPolicySampleCollector
+from ray.rllib.evaluation.sample_collector import _SampleCollector
+from ray.rllib.policy.policy import Policy
+from ray.rllib.policy.sample_batch import MultiAgentBatch
+from ray.rllib.utils import force_list
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.debug import summarize
+from ray.rllib.utils.types import AgentID, EnvID, EpisodeID, PolicyID, \
+    TensorType
+from ray.util.debug import log_once
+
+logger = logging.getLogger(__name__)
+
+
+class _MultiAgentSampleCollector(_SampleCollector):
+    """Builds SampleBatches for each policy (and agent) in a multi-agent env.
+
+    Note: This is an experimental class only used when
+    `config._use_trajectory_view_api` = True.
+    Once `_use_trajectory_view_api` becomes the default in configs:
+    This class will deprecate the `SampleBatchBuilder` class.
+
+    Input data is collected in central per-policy buffers, which
+    efficiently pre-allocate memory (over n timesteps) and re-use the same
+    memory even for succeeding agents and episodes.
+    Input_dicts for action computations, SampleBatches for postprocessing, and
+    train_batch dicts are - if possible - created from the central per-policy
+    buffers via views to avoid copying of data).
+    """
+
+    def __init__(self,
+                 policy_map: Dict[PolicyID, Policy],
+                 callbacks: "DefaultCallbacks",
+                 # TODO: (sven) make `num_agents` flexibly grow in size.
```

Comment:
We'll need this before we can turn this on by default, correct?

Reply:
Actually, right now, we assume that the number of agents may change all the time and therefore I haven't built in any more efficient handling of creating the forward pass batch (meaning, we still copy for that purpose simply by doing single-indexing into the main buffer (which causes copies)). I was thinking of automating detecting, whether agents are always the same (and step at the same time) or not. Then we can avoid even these copies and further make things faster.